File: Android/src/main/java/com/example/Android/Services/MovieService.java
package com.example.Android.Services;

import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.Android.DTO.Requests.CreateMovieRequest;
import com.example.Android.DTO.Requests.UpdateMovieRequest;
import com.example.Android.DTO.Responses.MovieResponse;
import com.example.Android.DTO.Responses.MoviesListResponse;
import com.example.Android.Models.Genre;
import com.example.Android.Models.Movie;
import com.example.Android.Repositories.GenreRepository;
import com.example.Android.Repositories.MovieRepository;
import com.example.Android.Repositories.ShowtimeRepository;

import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = lombok.AccessLevel.PRIVATE, makeFinal = true)
@Slf4j
public class MovieService {
    MovieRepository movieRepository;
    GenreRepository genreRepository;
    ShowtimeRepository showtimeRepository;
    
    public MoviesListResponse getAllMovies() {
        log.info("Fetching all movies");
        List<Movie> movies = movieRepository.findAllByOrderByIdDesc();
        
        List<MovieResponse> movieResponses = movies.stream()
                .map(this::convertToMovieResponse)
                .collect(Collectors.toList());
        
        log.info("Found {} movies", movieResponses.size());
        return MoviesListResponse.builder()
                .movies(movieResponses)
                .totalCount(movieResponses.size())
                .message("Movies retrieved successfully")
                .build();
    }
    
    public Optional<MovieResponse> getMovieById(Long id) {
        log.info("Fetching movie with ID: {}", id);
        Optional<Movie> movie = movieRepository.findById(id);
        
        if (movie.isPresent()) {
            log.info("Movie found: {}", movie.get().getTitle());
            return Optional.of(convertToMovieResponse(movie.get()));
        } else {
            log.warn("Movie not found with ID: {}", id);
            return Optional.empty();
        }
    }
    
    public MoviesListResponse searchMovies(String keyword) {
        log.info("Searching movies with keyword: {}", keyword);
        List<Movie> movies = movieRepository.findByKeyword(keyword);
        
        List<MovieResponse> movieResponses = movies.stream()
                .map(this::convertToMovieResponse)
                .collect(Collectors.toList());
        
        log.info("Found {} movies matching keyword: {}", movieResponses.size(), keyword);
        return MoviesListResponse.builder()
                .movies(movieResponses)
                .totalCount(movieResponses.size())
                .message("Search results for: " + keyword)
                .build();
    }
    
    public MoviesListResponse getMoviesByGenre(String genre) {
        log.info("Fetching movies by genre: {}", genre);
        List<Movie> movies = movieRepository.findByGenreName(genre);
        
        List<MovieResponse> movieResponses = movies.stream()
                .map(this::convertToMovieResponse)
                .collect(Collectors.toList());
        
        log.info("Found {} movies in genre: {}", movieResponses.size(), genre);
        return MoviesListResponse.builder()
                .movies(movieResponses)
                .totalCount(movieResponses.size())
                .message("Movies in genre: " + genre)
                .build();
    }
    
    public MoviesListResponse getHighRatedMovies(Double minRating) {
        log.info("Fetching high-rated movies with rating >= {}", minRating);
        List<Movie> movies = movieRepository.findByRatingGreaterThanEqual(minRating);
        
        List<MovieResponse> movieResponses = movies.stream()
                .map(this::convertToMovieResponse)
                .collect(Collectors.toList());
        
        log.info("Found {} high-rated movies", movieResponses.size());
        return MoviesListResponse.builder()
                .movies(movieResponses)
                .totalCount(movieResponses.size())
                .message("Featured movies with rating >= " + minRating)
                .build();
    }
    
    // ============ ADMIN CRUD OPERATIONS ============
    
    public MovieResponse createMovie(CreateMovieRequest request) {
        log.info("Creating new movie: {}", request.getTitle());
        
        // Create movie entity
        Movie movie = Movie.builder()
                .title(request.getTitle())
                .description(request.getDescription())
                .genre(request.getGenre())
                .duration(request.getDuration())
                .rating(request.getRating())
                .posterUrl(request.getPosterUrl())
                .backdropUrl(request.getBackdropUrl())
                .trailerUrl(request.getTrailerUrl())
                .releaseDate(request.getReleaseDate())
                .status(request.getStatus())
                .genres(new HashSet<>())
                .build();
        
        // Add genres if provided
        if (request.getGenreIds() != null && !request.getGenreIds().isEmpty()) {
            Set<Genre> genres = new HashSet<>();
            for (Long genreId : request.getGenreIds()) {
                genreRepository.findById(genreId).ifPresent(genres::add);
            }
            movie.setGenres(genres);
        }
        
        Movie savedMovie = movieRepository.save(movie);
        log.info("Movie created successfully with ID: {}", savedMovie.getId());
        
        return convertToMovieResponse(savedMovie);
    }
    
    public MovieResponse updateMovie(Long id, UpdateMovieRequest request) {
        log.info("Updating movie with ID: {}", id);
        
        Movie movie = movieRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Movie not found with ID: " + id));
        
        // Update fields if provided
        if (request.getTitle() != null) movie.setTitle(request.getTitle());
        if (request.getDescription() != null) movie.setDescription(request.getDescription());
        if (request.getGenre() != null) movie.setGenre(request.getGenre());
        if (request.getDuration() != null) movie.setDuration(request.getDuration());
        if (request.getRating() != null) movie.setRating(request.getRating());
        if (request.getPosterUrl() != null) movie.setPosterUrl(request.getPosterUrl());
        if (request.getBackdropUrl() != null) movie.setBackdropUrl(request.getBackdropUrl());
        if (request.getTrailerUrl() != null) movie.setTrailerUrl(request.getTrailerUrl());
        if (request.getReleaseDate() != null) movie.setReleaseDate(request.getReleaseDate());
        if (request.getStatus() != null) movie.setStatus(request.getStatus());
        
        // Update genres if provided
        if (request.getGenreIds() != null) {
            Set<Genre> genres = new HashSet<>();
            for (Long genreId : request.getGenreIds()) {
                genreRepository.findById(genreId).ifPresent(genres::add);
            }
            movie.setGenres(genres);
        }
        
        Movie updatedMovie = movieRepository.save(movie);
        log.info("Movie updated successfully: {}", updatedMovie.getTitle());
        
        return convertToMovieResponse(updatedMovie);
    }
    
    @Transactional
    public void deleteMovie(Long id) {
        log.info("Deleting movie with ID: {}", id);
        
        Movie movie = movieRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Movie not found with ID: " + id));

        List<com.example.Android.Models.Showtime> showtimes = showtimeRepository.findByMovieId(id);
        if (!showtimes.isEmpty()) {
            showtimeRepository.deleteAll(showtimes);
        }
        
        movieRepository.delete(movie);
        log.info("Movie deleted successfully: {}", movie.getTitle());
    }
    
    public MoviesListResponse getMoviesByStatus(String status) {
        log.info("Fetching movies with status: {}", status);
        List<Movie> movies = movieRepository.findByStatus(status);
        
        List<MovieResponse> movieResponses = movies.stream()
                .map(this::convertToMovieResponse)
                .collect(Collectors.toList());
        
        log.info("Found {} movies with status: {}", movieResponses.size(), status);
        return MoviesListResponse.builder()
                .movies(movieResponses)
                .totalCount(movieResponses.size())
                .message("Movies with status: " + status)
                .build();
    }
    
    private MovieResponse convertToMovieResponse(Movie movie) {
        List<String> genreNames = movie.getGenres().stream()
                .map(genre -> genre.getName())
                .collect(Collectors.toList());
                
        return MovieResponse.builder()
                .id(movie.getId())
                .title(movie.getTitle())
                .description(movie.getDescription())
                .genre(movie.getGenre())
                .genres(genreNames)
                .duration(movie.getDuration())
                .rating(movie.getRating())
                .posterUrl(movie.getPosterUrl())
                .backdropUrl(movie.getBackdropUrl())
                .trailerUrl(movie.getTrailerUrl())
                .releaseDate(movie.getReleaseDate())
                .status(movie.getStatus())
                .createdAt(movie.getCreatedAt())
                .build();
    }
}

File: mobile/MovieBookingMobile/app/services/movieService.js
import { apiClient } from './apiClient';

export const movieService = {
  getAllMovies: async (params = {}) => {
    try {
      console.log('movieService.getAllMovies called with params:', params);
      const response = await apiClient.get('/api/movies', { params });
      console.log('Movies API response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error fetching movies:', error);
      throw error;
    }
  },

  getMovieById: async (id) => {
    try {
      const response = await apiClient.get(`/api/movies/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching movie ${id}:`, error);
      throw error;
    }
  },

  searchMovies: async (keyword) => {
    try {
      const response = await apiClient.get('/api/movies/search', {
        params: { keyword }
      });
      return response.data;
    } catch (error) {
      console.error('Error searching movies:', error);
      throw error;
    }
  },

  getMoviesByGenre: async (genre) => {
    try {
      const response = await apiClient.get(`/api/movies/genre/${genre}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching movies by genre ${genre}:`, error);
      throw error;
    }
  },

  getFeaturedMovies: async () => {
    try {
      const response = await apiClient.get('/api/movies/featured');
      return response.data;
    } catch (error) {
      console.error('Error fetching featured movies:', error);
      throw error;
    }
  },

  getRecommendationGenres: async () => {
    try {
      const response = await apiClient.get('/api/recommendations/genres');
      return response.data;
    } catch (error) {
      console.error('Error fetching recommendation genres:', error);
      throw error;
    }
  },

  selectRecommendationGenres: async (genreIds = []) => {
    try {
      const body = {
        genres: genreIds.map((id) => ({ id })),
      };
      const response = await apiClient.post('/api/recommendations/select/genres', body);
      return response.data;
    } catch (error) {
      console.error('Error saving selected recommendation genres:', error);
      throw error;
    }
  },

  getRecommendedMovies: async (limit = 10) => {
    try {
      const response = await apiClient.get('/api/recommendations/movies', {
        params: { limit },
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching recommended movies:', error);
      throw error;
    }
  },

  selectMovieForRecommendation: async (movieId) => {
    try {
      const response = await apiClient.post('/api/movies/select', { movieId });
      return response.data;
    } catch (error) {
      console.error('Error saving movie selection:', error);
      throw error;
    }
  },

  // ============ ADMIN CRUD METHODS ============
  
  createMovie: async (movieData) => {
    try {
      const response = await apiClient.post('/api/admin/movies', movieData);
      return response.data;
    } catch (error) {
      console.error('Error creating movie:', error);
      throw error;
    }
  },

  updateMovie: async (id, movieData) => {
    try {
      const response = await apiClient.put(`/api/admin/movies/${id}`, movieData);
      return response.data;
    } catch (error) {
      console.error(`Error updating movie ${id}:`, error);
      throw error;
    }
  },

  deleteMovie: async (id) => {
    try {
      await apiClient.delete(`/api/admin/movies/${id}`);
    } catch (error) {
      console.error(`Error deleting movie ${id}:`, error);
      throw error;
    }
  },

  getMoviesByStatus: async (status) => {
    try {
      const response = await apiClient.get(`/api/admin/movies/status/${status}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching movies by status ${status}:`, error);
      throw error;
    }
  }
};

export const genreService = {
  getAllGenres: async () => {
    try {
      const response = await apiClient.get('/api/genres');
      return response.data;
    } catch (error) {
      console.error('Error fetching genres:', error);
      throw error;
    }
  }
};

